python/unitytrainers/ppo/models.py

```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
         update_reward = tf.assign(last_reward, new_reward)
         return last_reward, new_reward, update_reward
 
+    def create_curiosity_encoders(self):
+        """
+        Creates state encoders for current and future observations.
+        Used for implementation of ﻿Curiosity-driven Exploration by Self-supervised Prediction
+        See https://arxiv.org/abs/1705.05363 for more details.
+        :return: current and future state encoder tensors.
+        """
+        inverse_input_list = []
+        encoded_state_list = []
+        encoded_next_state_list = []
+
+        if self.v_size > 0:
+            self.next_visual_in = []
+            visual_encoders = []
+            next_visual_encoders = []
+            for i in range(self.v_size):
+                # Create input ops for next (t+1) visual observations.
+                next_visual_input = self.create_visual_input(self.brain.camera_resolutions[i],
+                                                             name="next_visual_observation_" + str(i))
+                self.next_visual_in.append(next_visual_input)
+
+                # Create the encoder ops for current and next visual input. Not that these encoders are siamese.
+                encoded_visual = self.create_visual_observation_encoder(self.visual_in[i], 128,
+                                                                        self.swish, 1, "visual_obs_encoder", False)
+                encoded_next_visual = self.create_visual_observation_encoder(self.next_visual_in[i], 128,
+                                                                             self.swish, 1, "visual_obs_encoder", True)
+                visual_encoders.append(encoded_visual)
+                next_visual_encoders.append(encoded_next_visual)
+
+            hidden_visual = tf.concat(visual_encoders, axis=1)
+            hidden_next_visual = tf.concat(next_visual_encoders, axis=1)
+            inverse_input_list.append(hidden_visual)
+            inverse_input_list.append(hidden_next_visual)
+            encoded_state_list.append(hidden_visual)
+            encoded_next_state_list.append(hidden_next_visual)
+
+        if self.o_size > 0:
```

Comment:
I think we have a better way to check for continuous vector observation than than. It might cause problems in scenarios where the vector observation is a continuous single float.

Reply:
What would you suggest?